Support for user-defined scripts to customize pipeline jobs

scottwittenburg opened the pull request:
Allow users to specify whatever `before_script` and `after_script` they want in their `spack.yaml`, on a per-runner basis.  Users can also specify their own `script`, but if they don't, we'll generate the standard `spack ci rebuild` script for them.

For a more in-depth discussion of this proposed change, see #17164, or take a look at the test and documentation added by this change.

scottwittenburg committed:
Provide your own script, before_script, and after_script

scottwittenburg commented:
I think this is way cleaner than what I proposed in #17119, but it's still to be seen whether it gives everyone what they need.

scottwittenburg committed:
Support variable interpolation at pipeline generation time

scottwittenburg committed:
Add test for variable interpolation and scripts

scottwittenburg committed:
Update pipeline documentation to describe user-provided scripts

scottwittenburg committed:
Make sure each develop pipeline tests associated commit

scottwittenburg committed:
Allow overridable global runner attributes

scottwittenburg commented:
I ran a small [pipeline](https://gitlab.spack.io/spack/e4s/pipelines/4421) with this change, just to demonstrate how it could be used, and to show that it actually works 😄

scottwittenburg committed:
Document pipeline keys which can be global but overridden

Update pipelines documentation to describe how 'tags', 'variables',
'image', 'before_script', 'script', and 'after_script' can be
supplied at the top level, to be used by any of the runner mappings,
and also overridden by any of the runner mappings.

Also show an example of capturing the custom spack SHA at pipeline
generation time, so all jobs are sure to run with the same version
of spack, as a means to illustrate the $env:VARIABLE_NAME syntax.

opadron commented on lib/spack/docs/pipelines.rst:
```diff
@@ -245,7 +280,21 @@ runners known to the gitlab instance.  For Docker executor type runners, the
+``$env:<var_value>``, then that variable will be inserted with ``var_value``
+first sampled from the environment during job generation.  This can be useful
+for propagating generation-time variables which normally would not be available
+child build jobs.
```
Possible typo?  "...which normally would not be available *to* child build jobs. ..."

opadron commented on lib/spack/docs/pipelines.rst:
```diff
@@ -245,7 +280,21 @@ runners known to the gitlab instance.  For Docker executor type runners, the
+``before_script``, and ``after_script`` sections to be applied to every job
+scheduled on that runner.  This allows users to do any custom preparation or
+cleanup tasks that fit their particular workflow, as well as completely
+customize the rebuilding of a spec if they so choose.  We will never generate
+a ``before_script`` or ``after_script`` for jobs, but if you do not provide
+a custom ``script``, we will generate one for you that invokes
+``spack ci rebuild``.
```
Two points:

 1. Consider rephrasing to remove the use of self-referential pronouns.  Consider passive voice.  (e.g.: "Neither before_script nor after_script sections are generated by default ...")

 1. You might not want to document the fact that there are no before or after scripts generated, lest they become part of the de-facto API surface.  Maybe just say that a user-provided section will never be overridden by anything we automatically generate.  This language gives us an out of we ever need to generate a default `before_script` or `after_script` section in the future.

tldahlgren replied:
Is passive void desirable?

scottwittenburg replied:
Maybe instead of using the passive voice, I can replace `we` with `spack` in most instances?

opadron commented on lib/spack/docs/pipelines.rst:
```diff
@@ -368,13 +417,89 @@ Using a custom spack in your pipeline
+could use the GitLab user interface to create CI environment variables
+containing the url and branch or tag you want to clone (calling them, for
+example, ``SPACK_REPO`` and ``SPACK_REF``), then use those in a custom shell
+script invoked both from your pipeline generation job, as well in your rebuild
```
Possible typo: "... as well *as in your rebuild ..."

opadron commented on lib/spack/docs/pipelines.rst:
```diff
@@ -383,25 +508,58 @@ a custom spack and make sure the generated rebuild jobs will clone it too:
+the generated yaml of the job as a variable we can get when the child job
+runs (``CAPTURED_SPACK_SHA``).  Those variables can actually have the same name,
+but to avoid confusion, we've made them distinct here.  Now all your child
+jobs will clone the same SHA of spack used to generate the pipeline, no matter
+whether more commits are pushed to the branch during that time.
```
Phrasing feels a little strange.  What do you think of "... of spack used to generate the pipeline, *even *if more commits are pushed ..."

opadron commented on lib/spack/spack/ci.py:
```diff
@@ -40,6 +42,8 @@
 spack_gpg = SpackCommand('gpg')
 spack_compiler = SpackCommand('compiler')
 
+runner_var_regex = re.compile('\\$env:(.+)$')
```
Would this regex catch something like this?

`my_var: EXTRA_GRABAGE$env:MY_VAR`

Is this desired, or should this match only on values that also *start* with `$env`?

If the latter, consider adding `^` to the regex (i.e.: `'^\\$env:(.+)$'`)

scottwittenburg replied:
Hmmm, seems like allowing it to replace a substring, as you showed in your extra garbage example, could be useful.  In which case I don't want to bind to the beginning of the string, but I would need to make sure to replace only the matched portion with the interpolated variable, rather than the whole string.

I'll add this change, but keep it as a separate commit.  If people have strong objections, I can always remove it.

opadron commented on lib/spack/spack/ci.py:
```diff
@@ -592,6 +619,13 @@ def generate_gitlab_ci_yaml(env, print_summary, output_file,
                 if 'variables' in runner_attribs:
                     variables.update(runner_attribs['variables'])
+                    for name, value in variables.items():
+                        m = runner_var_regex.search(value)
```
As an alternative to my above comment about the regex:

Consider using `runner_var_regex.match(value)` instead of `search()`.  This change should also have the effect of requiring that the match occur at the very beginning of `value`.

opadron commented on lib/spack/spack/test/cmd/ci.py:
```diff
@@ -391,6 +390,88 @@ def test_ci_generate_with_cdash_token(tmpdir, mutable_mock_env_path,
+def test_ci_generate_with_script_and_variables(tmpdir, mutable_mock_env_path,
+                                               env_deactivate, install_mockery,
+                                               mock_packages):
+    """Make sure we it doesn't break if we configure cdash"""
```
Possible typo?

scottwittenburg replied:
Yes, copy-paste error, thanks.

opadron reviewed (approved):
Looks good.  See comments for suggested changes. :+1:

tldahlgren reviewed (commented):
I really like the flexibility of this change.  

I ran the tests and all but two passed on a linux machine.  The two failing tests are unrelated to this PR as they've failed for me on `develop`.

I want to review the changes more closely before I complete my review.

tldahlgren commented on lib/spack/docs/pipelines.rst:
```diff
@@ -245,7 +280,21 @@ runners known to the gitlab instance.  For Docker executor type runners, the
+``before_script``, and ``after_script`` sections to be applied to every job
+scheduled on that runner.  This allows users to do any custom preparation or
+cleanup tasks that fit their particular workflow, as well as completely
+customize the rebuilding of a spec if they so choose.  We will never generate
```
IMO `never` should be changed to `not` or some similar wording to indicate that this statement applies at _this_ time.

scottwittenburg replied:
Makes sense, will do.

tldahlgren commented on lib/spack/docs/pipelines.rst:
```diff
@@ -368,13 +417,89 @@ Using a custom spack in your pipeline
+             after_script:
+               - rm -rf ./spack
+
+Now all the generated rebuild jobs will use the same shell script to clone
```
Suggest adding `of` in the middle of `all the`.

tldahlgren commented on lib/spack/docs/pipelines.rst:
```diff
@@ -383,25 +508,58 @@ a custom spack and make sure the generated rebuild jobs will clone it too:
+               - spack-kube
+             image: spack/ubuntu-bionic
+             variables:
+               CAPTURED_SPACK_SHA: $env:CURRENT_SPACK_SHA
```
I know this has been discussed in meetings, but I **wish** there was a cleaner way to indicate deferral of environment variable evaluation than using `$env:`.

scottwittenburg replied:
What part of it do you think is tripping people up?  Is it simply a naming issue?

scottwittenburg replied:
@tgamblin ☝️

tldahlgren commented on lib/spack/spack/ci.py:
```diff
@@ -421,12 +425,53 @@ def spec_matches(spec, match_string):
+                    copy_attributes(overridable_attrs,
+                                    ci_mapping['runner-attributes'],
+                                    runner_attributes)
+                return runner_attributes
```
Is this `return` necessary?

scottwittenburg replied:
The point of the return at this point is to make sure we only process the first match, since we are supposed to map the spec to the first mapping that matches.  We definitely do not want to let attributes build up over several matches either.  Let me know if you see some problem with this I may have missed.

tldahlgren commented on lib/spack/docs/pipelines.rst:
```diff
@@ -189,15 +197,37 @@ corresponds to a known gitlab runner, where the ``match`` section is used
-provide this option is ``False``).  The ``enable-debug-messages`` key takes a boolean
+provide this option is ``False``).
+
+The ``enable-debug-messages`` key takes a boolean
```
Support for this setting appears to have been removed so the documentation here should also be removed.

scottwittenburg replied:
Yes, it has been removed, thanks for catching this.

tldahlgren reviewed (changes requested):
LGTM.  I have a few questions, suggestions, and changes.  The most important comment is ensuring consistency between the documentation and support for `enable-debug-messages`.

tldahlgren commented:
@adrienbernede Will this PR address your needs?

scottwittenburg commented:
> LGTM. I have a few questions, suggestions, and changes. The most important comment is ensuring consistency between the documentation and support for `enable-debug-messages`.

Sounds good, I'm taking a pass now to address the issues raised by you and @opadron.  Thanks for your review 😄

tgamblin reviewed (changes requested):
@scottwittenburg: can you refresh my memory on why it has to be `$env:name` vs just `$name`?  If anything I think the name doesn't convey why the prefix is needed.  I think we should fix it -- @tldahlgren finds it odd too.

Is the issue that `$name` is evaluated in `spack.yaml` and `$env:name` is evaluated in `.gitlab-ci.yml`?  Or do I have that backwards?

scottwittenburg commented:
> @scottwittenburg: can you refresh my memory on why it has to be `$env:name` vs just `$name`? If anything I think the name doesn't convey why the prefix is needed. I think we should fix it -- @tldahlgren finds it odd too.

Anything in the variables section of a gitlab mapping in the  `spack.yaml` would normally be simply copied, verbatim, into the generated build job.  So `$name` would end up in the generated job without any translation, and this would result in the gitlab interpolating the environment variable `name` when the rebuild job runs.

The aim of the `$env:name` syntax was to give us a way to have spack query the environment variable `name` when the pipeline is generated, so that the interpolated variable value is written into the generated job, as sometimes we know an environment variable which is available during pipeline generation will no longer be there when the job runs.

> Is the issue that `$name` is evaluated in `spack.yaml` and `$env:name` is evaluated in `.gitlab-ci.yml`? Or do I have that backwards?

I think you have it backwards, if I understand your question.  Gitlab would use the dollar to treat it as an environment variable when it runs the job.  Sometimes that's what we want, but sometimes it's useful to have spack interpolate the variable during pipeline generation.

scottwittenburg committed:
Remove :<name> interpolation, add SPACK_VERSION variables

Also fix issues with documentation to reflect changes

scottwittenburg commented:
Link to [pipeline](https://gitlab.spack.io/spack/e4s/pipelines/4895) run after review changes, including removal of `$env:<VAR_NAME>` syntax.

scottwittenburg commented:
@opadron @tldahlgren I think I've addressed all your suggestions, thanks for your review.  I think this could be ready to go now, though we may want to synchronize merging it with a change to `gitlab.spack.io/spack/e4s` so the transition in workflow is as smooth as possible.

fyi @tgamblin @becker33

scottwittenburg committed:
Do not assume we sit in the directory where the env file lives.

scottwittenburg commented:
@adrienbernede Thanks for trying this out!  I added a commit ([7c1c26c](https://github.com/spack/spack/pull/17129/commits/7c1c26ca41b06f3fb2ba13612f5849929faf9145)) fixing the issue you found with running `spack ci rebuild` from a directory that did not contain the `spack.yaml` file.  I ran a pipeline to test it, you can see the results [here](https://gitlab.spack.io/spack/e4s/pipelines/5254).

shahzebsiddiqui commented:
So i am kinda dependent on this PR however i am not excited about adding a `script` section in `spack.yaml`. Can't we get `SPACK_SIGNING_KEY` working and let spack figure out how to get spack to generate the commands properly. The issue with Cori is gpg key was not added in child job. I kinda liked having spack gpg trust in parent job that trusts gpg key in parent job. I would want this command be added in child job or have a way to propagate this into child job.

I think user won't understand what commands are generated by spack and what user will specify. Furthermore, i feel that user wont understand what they are suppose to do in the `script` section considering spack is regenerating a pipeline.yml. It just makes this process more confusing.

shahzebsiddiqui commented:
I wanted to point out this PR is a blocker for us to use spack upstream. We kinda need this need feature in spack.yaml so for interim we are using Luke fork https://github.com/eugeneswalker/spack/tree/ci-2 to build our pipeline.

tldahlgren reviewed (approved):
LGTM.
